### PR TITLE
Docs: code-first federation

### PR DESCRIPTION
**Pull request by tuxmachine**
A rough, first draft
Examples need to be tested and some extra text would be nice

## PR Checklist
- [X] The commit message follows our guidelines: https://github.com/nestjs/docs.nestjs.com/blob/master/CONTRIBUTING.md

## PR Type
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[X] Other... Please describe: Docs
```

### CODE REVIEW

**Comment by kamilmysliwiec** [comment]
Looks amazing so far @tuxmachine! Thank you

File: content/graphql/federation.md [hunk 1]
```diff
@@ -61,21 +63,88 @@ Finally, we hook everything up in a module together with a `GraphQLFederationMod
...
+    }),
+  ],
+  providers: [UserResolver],
 })
 export class AppModule {}
 ```
 
 #### Federated example: Posts
 
+Our Post microservice serves aggregated posts via a `getPosts` query, but also extends our User type with `user.posts`
```
Comment (johnbiundo) [inline content/graphql/federation.md:144]:
```suggestion
Our Post service serves aggregated posts via a `getPosts` query, but also extends our `User` type with `user.posts`
```

File: content/graphql/federation.md [hunk 2]
```diff
@@ -137,11 +206,134 @@ import { PostsResolvers } from './posts.resolvers';
 export class AppModule {}
 ```
 
+##### Code first
+
+We will need to create a class representing our User entity. Even though it lives in another microservice, we will be using and extending it. Note the `@extends` and `@external` directives
```
Comment (johnbiundo) [inline content/graphql/federation.md:211]:
```suggestion
We will need to create a class representing our `User` entity. Even though it lives in another service, we will be using and extending it. Note the `@extends` and `@external` directives.
```

File: content/graphql/federation.md [hunk 3]
```diff
@@ -137,11 +206,134 @@ import { PostsResolvers } from './posts.resolvers';
...
+  @Field(type => [Post])
+  public posts: Post[];
+
+  constructor(post: Partial<User>) {
+    Object.assign(this, post);
+  }
+}
+```
+
+We create the resolver for our extension on the User entity as follows:
```
Comment (johnbiundo) [inline content/graphql/federation.md:234]:
```suggestion
We create the resolver for our extension on the `User` entity as follows:
```

File: content/graphql/federation.md [hunk 4]
```diff
@@ -137,11 +206,134 @@ import { PostsResolvers } from './posts.resolvers';
...
+  constructor(private readonly postService: PostService) {}
+
+  @ResolveField(of => [Post])
+  public posts(@Parent() user: User) {
+    return this.postService.forAuthor(user.id);
+  }
+}
+```
+
+We also need to create our Post entity
```
Comment (johnbiundo) [inline content/graphql/federation.md:253]:
```suggestion
We also need to create our `Post` entity:
```

File: content/graphql/federation.md [hunk 5]
```diff
@@ -137,11 +206,134 @@ import { PostsResolvers } from './posts.resolvers';
...
+  @Field(type => User)
+  public user: User;
+
+  constructor(post: Partial<Post>) {
+    Object.assign(this, post);
+  }
+}
+```
+
+And its resolver
```
Comment (johnbiundo) [inline content/graphql/federation.md:280]:
```suggestion
And its resolver:
```

File: content/graphql/federation.md [hunk 6]
```diff
@@ -137,11 +206,134 @@ import { PostsResolvers } from './posts.resolvers';
...
+  }
+
+  @ResolveField(of => User)
+  public user(@Parent() post) {
+    return { __typename: 'User', id: post.userId };
+  }
+}
+```
+
+And finally tie it together in a module. Note the schema build options, where we specify that User is an outside type.
```
Comment (johnbiundo) [inline content/graphql/federation.md:309]:
```suggestion
And finally tie it together in a module. Note the schema build options, where we specify that `User` is an outside type.
```

**Comment by brandon-leapyear** [comment]
Hello! First of all, thanks for your hard work here. Can't wait to get more code-first docs on the website.

My team is trying to migrate an Apollo Server graphql server to a NestJS graphql server, so we're using Apollo Federation as a separate service to combine the two. I'm trying to follow the instructions here, but I'm still getting `Cannot query field "_entities" on type "Query"` and `Unknown type "_Any"`. I notice that the Apollo Federation spec [here](https://www.apollographql.com/docs/apollo-server/federation/federation-spec/) says that any federation implementation needs to add these things to the schema. Is there a way to get the code-first approach to include these federation requirements when generating the schema?

**Comment by Shnikts** [comment]
Hello,
I have been trying to make a federated monorepo project with code first style following this documentation, but I am having trouble making the gateway work. Is there any chance that there is an example project that I look at?

**Comment by tuxmachine** [comment]
Sorry about the delay on finishing the docs, just had a first kid and it's taking a little more time than expected. 

@Shnikts You could have a look at the [tests](https://github.com/nestjs/graphql/tree/master/tests/graphql-federation/gateway). There's nothing code-first about the gateway though, so the existing schema-first docs apply.

**Comment by kamilmysliwiec** [comment]
@tuxmachine is this ready for rereview by chance? :)

**Comment by StevenLangbroek** [comment]
I followed the docs and got it working 🥳

**Comment by iamolegga** [comment]
> I followed the docs and got it working 🥳

The same for me, works like a charm. Thanks @tuxmachine 👍

**Comment by tuxmachine** [comment]
Sorry it took a while. I've tested the examples and fixed some minor stuff. You can find the code-samples collected into a working demo here: https://github.com/tuxmachine/nestjs-federation-example

@kamilmysliwiec Yes it's ready for re-review

**Comment by melkir** [comment]
Thank you @tuxmachine, I've successfully implemented my own federated schema based on yours! 😁
A small difference inside my implementation is that I'm using a gateway in order to stitch the services. Here's my demo: https://github.com/melkir/nestjs-federation-example

**Comment by mikhailbalin** [comment]
@melkir That's not working for me:
```
Error checking for changes to service definitions: Couldn't load service definitions for "users" at http://localhost:3001/graphql: request to http://localhost:3001/graphql failed, reason: connect ECONNREFUSED 127.0.0.1:3001
This data graph is missing a valid configuration. Couldn't load service definitions for "users" at http://localhost:3001/graphql: request to http://localhost:3001/graphql failed, reason: connect ECONNREFUSED 127.0.0.1:3001
```

**Comment by kamilmysliwiec** [comment]
Thanks 🔥

**Pull request merged**